File: sdk/go/device_registry.go
package sdk

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"regexp"

	"github.com/hyperledger/fabric-gateway/pkg/client"
	"github.com/nexus-lab/iot-service-blockchain/common"
)

// DeviceEvent an event emitted by the device registry contract notifying a device update
type DeviceEvent struct {
	// Action name of the action performed on the device
	Action string

	// OrganizationId organization ID of the device
	OrganizationId string

	// DeviceId ID of the device
	DeviceId string

	// Payload custom event payload
	Payload interface{}
}

// DeviceRegistryInterface core utilities for managing devices on the ledger
type DeviceRegistryInterface interface {
	// Register create or update a device in the ledger
	Register(device *common.Device) error

	// Get return a device by its organization ID and device ID
	Get(organizationId string, deviceId string) (*common.Device, error)

	// GetAll return a list of devices by their organization ID
	GetAll(organizationId string) ([]*common.Device, error)

	// Deregister remove a device from the ledger
	Deregister(device *common.Device) error

	// RegisterEvent registers for device registry events
	RegisterEvent(options ...client.ChaincodeEventsOption) (<-chan *DeviceEvent, context.CancelFunc, error)
}

// DeviceRegistry core utilities for managing devices on the ledger
type DeviceRegistry struct {
	contract ContractInterface
}

// Register create or update a device in the ledger
func (r *DeviceRegistry) Register(device *common.Device) error {
	if device == nil {
		return fmt.Errorf("cannot register an empty device")
	}

	data, err := device.Serialize()
	if err != nil {
		return err
	}

	_, err = r.contract.SubmitTransaction("Register", string(data))
	return err
}

// Get return a device by its organization ID and device ID
func (r *DeviceRegistry) Get(organizationId string, deviceId string) (*common.Device, error) {
	data, err := r.contract.SubmitTransaction("Get", organizationId, deviceId)
	if err != nil {
		return nil, err
	}

	return common.DeserializeDevice(data)
}

// GetAll return a list of devices by their organization ID
func (r *DeviceRegistry) GetAll(organizationId string) ([]*common.Device, error) {
	data, err := r.contract.SubmitTransaction("GetAll", organizationId)
	if err != nil {
		return nil, err
	}

	results := make([]*common.Device, 0)
	if err = json.Unmarshal(data, &results); err != nil {
		return nil, err
	}

	return results, nil
}

// Deregister remove a device from the ledger
func (r *DeviceRegistry) Deregister(device *common.Device) error {
	if device == nil {
		return fmt.Errorf("cannot deregister an empty device")
	}

	data, err := device.Serialize()
	if err != nil {
		return err
	}

	_, err = r.contract.SubmitTransaction("Deregister", string(data))
	return err
}

// RegisterEvent registers for device registry events
func (r *DeviceRegistry) RegisterEvent(options ...client.ChaincodeEventsOption) (<-chan *DeviceEvent, context.CancelFunc, error) {
	dest := make(chan *DeviceEvent)
	source, cancel, err := r.contract.RegisterEvent(options...)
	pattern := regexp.MustCompile(`^device:\/\/(.+?)\/(.+?)\/(.+?)$`)

	go func() {
		defer close(dest)

		for event := range source {
			matches := pattern.FindStringSubmatch(event.EventName)
			if len(matches) != 4 {
				continue
			}

			deviceEvent := &DeviceEvent{
				OrganizationId: matches[1],
				DeviceId:       matches[2],
				Action:         matches[3],
			}

			if deviceEvent.Action == "register" || deviceEvent.Action == "deregister" {
				device, err := common.DeserializeDevice(event.Payload)
				if err != nil {
					log.Printf("bad device event payload %#v, action is %s\n", event.Payload, deviceEvent.Action)
					continue
				}
				deviceEvent.Payload = device
			} else {
				deviceEvent.Payload = event.Payload
			}

			dest <- deviceEvent
		}
	}()

	return dest, cancel, err
}

func createDeviceRegistry(network *client.Network, chaincodeId string) DeviceRegistryInterface {
	return &DeviceRegistry{
		contract: &Contract{
			network:      network,
			chaincodeId:  chaincodeId,
			contractName: "device_registry",
		},
	}
}

File: sdk/javascript/DeviceRegistry.ts
import {
  ChaincodeEventsOptions,
  CloseableAsyncIterable,
  Network,
} from '@hyperledger/fabric-gateway';
import { TextDecoder } from 'util';

import Contract, { ContractInterface } from './Contract';
import Device from './Device';

/**
 * Interface of the event emitted by the device registry contract notifying a device update
 */
export interface DeviceEvent {
  /**
   * Name of the action performed on the device
   */
  action: string;

  /**
   * Organization ID of the device
   */
  organizationId: string;

  /**
   * ID of the device
   */
  deviceId: string;

  /**
   * Custom event payload
   */
  payload: any;
}

/**
 * Interface of core utilities for managing devices on the ledger
 */
export interface DeviceRegistryInterface {
  /**
   * Create or update a device in the ledger
   *
   * @param device device to be created or updated
   */
  register(device: Device): Promise<void>;

  /**
   * Get a device by its organization ID and device ID
   *
   * @param organizationId device's organization ID
   * @param deviceId device's ID
   * @returns the device
   */
  get(organizationId: string, deviceId: string): Promise<Device>;

  /**
   * Get a list of devices by their organization ID
   *
   * @param organizationId devices' organization ID
   * @returns all devices of the organization
   */
  getAll(organizationId: string): Promise<Device[]>;

  /**
   * Remove a device from the ledger
   *
   * @param device the device to be removed
   */
  deregister(device: Device): Promise<void>;

  /**
   * Registers for device registry events
   *
   * @param options chaincode event options
   * @returns an async iterable of the events
   */
  registerEvent(options: ChaincodeEventsOptions): Promise<CloseableAsyncIterable<DeviceEvent>>;
}

/**
 * Core utilities for managing devices on the ledger
 */
export default class DeviceRegistry implements DeviceRegistryInterface {
  private utf8Decoder: TextDecoder;
  /**
   * @param contract smart contract
   */
  constructor(private contract: ContractInterface) {
    this.utf8Decoder = new TextDecoder();
  }

  async register(device: Device): Promise<void> {
    const serialized = device.serialize();
    await this.contract.submitTransaction('Register', serialized);
  }

  async get(organizationId: string, deviceId: string): Promise<Device> {
    const data = await this.contract.submitTransaction('Get', organizationId, deviceId);
    const serialized = this.utf8Decoder.decode(data);
    return Device.deserialize(serialized);
  }

  async getAll(organizationId: string): Promise<Device[]> {
    const data = await this.contract.submitTransaction('GetAll', organizationId);
    const serialized = this.utf8Decoder.decode(data);
    return (JSON.parse(serialized) as any[]).map(Device.fromObject);
  }

  async deregister(device: Device): Promise<void> {
    const serialized = device.serialize();
    await this.contract.submitTransaction('Deregister', serialized);
  }

  async registerEvent(
    options?: ChaincodeEventsOptions,
  ): Promise<CloseableAsyncIterable<DeviceEvent>> {
    const events = await this.contract.registerEvent(options);
    const pattern = /^device:\/\/(.+?)\/(.+?)\/(.+?)$/;
    const decoder = this.utf8Decoder;

    return {
      async *[Symbol.asyncIterator]() {
        for await (const event of events) {
          // can reuse pattern here since it has no global('g') flag
          const matches = pattern.exec(event.eventName);
          if (matches === null || matches.length != 4) {
            continue;
          }

          const deviceEvent: DeviceEvent = {
            organizationId: matches[1],
            deviceId: matches[2],
            action: matches[3],
            payload: null,
          };

          if (deviceEvent.action === 'register' || deviceEvent.action === 'deregister') {
            try {
              const device = Device.deserialize(decoder.decode(event.payload));
              deviceEvent.payload = device;
            } catch {
              console.error(
                `bad device event payload ${event.payload}, action is ${deviceEvent.action}`,
              );
              continue;
            }
          } else {
            deviceEvent.payload = event.payload;
          }

          yield deviceEvent;
        }
      },
      close() {
        events.close();
      },
    };
  }
}

export function createDeviceRegistry(network: Network, chaincodeId: string) {
  return new DeviceRegistry(new Contract(network, chaincodeId, 'device_registry'));
}
